securesystemslib/hsm.py

```diff
@@ -0,0 +1,446 @@
+      2. ECDSA
+
+    <Arguments>
+      data:
+        bytes over which the signature is to be calculated
+        'data' should be encoded/serialized before it is passed here.
+
+      private_key_handle:
+        element from the list returned by the get_private_key_objects()
+
+    <Exceptions>
+      securesystemslib.exceptions.UnsupportedAlgorithmError, when the
+      key type of the 'private_key_handle' is not supported
+
+    <Returns>
+      HEX string of the generated signature
+    """
+
+    mechanism = None
+    key_type = self.session.getAttributeValue(private_key_handle,
+        [PyKCS11.CKA_KEY_TYPE])[0]
+
+    if PyKCS11.CKK[key_type] == 'CKK_RSA':
+      mechanism = PyKCS11.RSA_PSS_Mechanism(RSA_PSS_MECH,
+          RSA_PSS_HASH_SHA256, RSA_PSS_MGF_SHA256, RSA_PSS_SALT_LENGTH)
+
+    elif PyKCS11.CKK[key_type] == 'CKK_EC' or PyKCS11.CKK[key_type] == 'CKK_ECDSA':
+      mechanism = PyKCS11.Mechanism(PyKCS11.CKM_ECDSA)
+
+    if mechanism is None:
+      raise securesystemslib.exceptions.UnsupportedAlgorithmError(
+        "The Key type " + repr(key_type) + " is currently not supported!")
+
+
+    signature = self.session.sign(private_key_handle, data,
+        mechanism)
+    return binascii.hexlify(bytes(signature)).decode('utf-8')
+
+
+
+
+
+  def verify_signature(self, signed_bytes, signature, public_key_handle):
+    """
+    <Purpose>
+      Verify that the corresponding private key of the public_key
+      generated 'signature' over 'signed_bytes'.
+
+      Supported Keys
+        1. RSA
+        2. ECDSA
+
+    <Arguments>
+      signed_bytes:
+        bytes over which the signature was calculated.
+        should be encoded/serialized before it is passed here.
+
+      signature:
+        HEX string generated by generate_signature()
+
+      private_key_handle:
+        element form the list returned by the get_private_key_objects()
+
+    <Exceptions>
+      securesystemslib.exceptions.UnsupportedAlgorithmError, when the
+      key type of the 'public_key_handle' is not supported
+
+    <Returns>
+      bool value for the correctness of the signature
+    """
+
+    # Convert HEX string to bytes
+    signature_bytes = binascii.unhexlify(signature)
+    mechanism = None
+    key_type = self.session.getAttributeValue(public_key_handle,
+        [PyKCS11.CKA_KEY_TYPE])[0]
+
+    if PyKCS11.CKK[key_type] == 'CKK_RSA':
+      mechanism = PyKCS11.RSA_PSS_Mechanism(RSA_PSS_MECH,
+          RSA_PSS_HASH_SHA256, RSA_PSS_MGF_SHA256, RSA_PSS_SALT_LENGTH)
+
+    elif PyKCS11.CKK[key_type] == 'CKK_EC' or PyKCS11.CKK[key_type] == 'CKK_ECDSA':
+      mechanism = PyKCS11.Mechanism(PyKCS11.CKM_ECDSA)
+
+    if mechanism is None:
+      raise securesystemslib.exceptions.UnsupportedAlgorithmError(
+          "The Key type " + repr(key_type) + " is currently not supported!")
+
+    result = self.session.verify(public_key_handle, signed_bytes, signature_bytes,
+        mechanism)
+
+    return result
+
+
+
+
+
+  def logout(self):
```

Comment:
Nit: Do `logout` and `close_session` form part of the public API, or are they just abstracted here for calling in `close`? If the latter, it might make sense to document them as private and prefix them with underscores (e.g. `_logout` and `_close_session`) to discourage misuse.